opentelemetry-sdk/src/opentelemetry/sdk/resources/__init__.py

```diff
@@ -371,6 +372,12 @@ def detect(self) -> "Resource":
         return Resource(resource_info)
 
 
+class ServiceInstanceIdResourceDetector(ResourceDetector):
+    # pylint: disable=no-self-use
+    def detect(self) -> "Resource":
+        return Resource({SERVICE_INSTANCE_ID: str(uuid.uuid4())})
```

Comment:
Is this returning something different at each invocation? Shouldn't we cache the result per process instead?

Reply:
> Is this returning something different at each invocation?

Yes.

> Shouldn't we cache the result per process instead?

It's definitely possible. However, in most languages and scenarios, detectors are called once and the resulting resource is used throughout the lifetime of the process, thus no caching on the level of resource detectors is needed. Do you see cases where caching would be beneficial?

Reply:
I'm worried about the pre-fork server (e.g. gunicorn) case where this need to return a different value for each worker process but the hopefully the same for the very same worker.

Reply:
I'll look into this scenario.

Reply:
i think this will also potentially result in different instance IDs per signal.
if no explicit resource is passed to a provider (i.e. `TracerProvider`, `MeterProvider`, `LoggerProvider`), the provider will create a new resource with  `Resource.create({})` which in turn calls all the resource detectors.

Reply:
Something simple like this may be enough?

```python
import os
import uuid

INSTANCE_ID_CACHE = {}


def get_instance_id() -> uuid.UUID:
    pid = os.getpid()
    return INSTANCE_ID_CACHE.setdefault(pid, uuid.uuid4())
```

Reply:
I changed the logic so there's now a stable (cached) instance id per process.